airbyte-webapp/src/hooks/services/FormChangeTracker/hooks.ts

```diff
@@ -0,0 +1,46 @@
+import { useCallback, useMemo } from "react";
+import { createGlobalState } from "react-use";
+import { useLocation } from "react-router-dom";
+import { uniqueId } from "lodash";
+
+import { FormChangeTrackerServiceApi } from "./types";
+
+export const useChangedFormsById = createGlobalState<Record<string, boolean>>({});
+
+export const useUniqueFormId = (formId?: string) => {
+  const location = useLocation();
+  return useMemo(
+    () => formId ?? `${location.pathname.toLowerCase().substring(1).replace(/\//gi, "_")}__${uniqueId("form_")}`,
+    [formId, location.pathname]
+  );
+};
+
+export const useFormChangeTrackerService = (): FormChangeTrackerServiceApi => {
+  const [changedFormsById, setChangedFormsById] = useChangedFormsById();
+
+  const clearAllFormChanges = useCallback(() => {
+    setChangedFormsById({});
+  }, [setChangedFormsById]);
+
+  const clearFormChange = useCallback(
+    (id: string) => {
+      setChangedFormsById({ ...changedFormsById, [id]: false });
+    },
+    [changedFormsById, setChangedFormsById]
+  );
+
+  const trackFormChange = useCallback(
+    (id: string, changed: boolean) => {
+      if (!!changedFormsById?.[id] !== changed) {
```

Comment:
Nitpick, this feels slightly clearer:
```suggestion
      if (Boolean(changedFormsById?.[id]) !== changed) {
```

Reply:
This casts the value correctly?

Reply:
Yes, I specified it was a nitpick as `!!changedFormsById?.[id] !== changed` looks a little strange in an if-statement.
I know it functions correctly but, and perhaps this is a personal thing, I could see this reading a little confusingly as `!!(changedFormsById?.[id] !== changed)`